cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java

```diff
@@ -0,0 +1,184 @@
+    // Any class that is not from cdap-api-spark or cdap-spark-core
+    // Any class that is not from Spark
+    if (API_CLASSES.contains(name) || (!name.startsWith("co.cask.cdap.api.spark.")
+        && !name.startsWith("co.cask.cdap.app.runtime.spark.")
+        && !name.startsWith("org.apache.spark."))) {
+      return super.loadClass(name, resolve);
+    }
+
+    // If the class is already loaded, return it
+    Class<?> cls = findLoadedClass(name);
+    if (cls != null) {
+      return cls;
+    }
+
+    // Define the class with this ClassLoader
+    try (InputStream is = getResourceAsStream(name.replace('.', '/') + ".class")) {
+      if (is == null) {
+        throw new ClassNotFoundException("Failed to find resource for class " + name);
+      }
+
+      if (name.equals(SPARK_CONTEXT_TYPE.getClassName())) {
+        // Define the SparkContext class by rewriting the constructor
+        cls = defineContext(SPARK_CONTEXT_TYPE, name, is);
+      } else if (name.equals(SPARK_STREAMING_CONTEXT_TYPE.getClassName())) {
+        // Define the StreamingContext class by rewriting the constructor
+        cls = defineContext(SPARK_STREAMING_CONTEXT_TYPE, name, is);
+      } else {
+        // Otherwise, just define it with this ClassLoader
+        byte[] byteCode = ByteStreams.toByteArray(is);
+        cls = defineClass(name, byteCode, 0, byteCode.length);
+      }
+
+      if (resolve) {
+        resolveClass(cls);
+      }
+      return cls;
+    } catch (IOException e) {
+      throw new ClassNotFoundException("Failed to read class definition for class " + name, e);
+    }
+  }
+
+  /**
+   * This method will define the class by rewriting the constructor to call
+   * {@link SparkContextCache#setContext(SparkContext)} or {@link SparkContextCache#setContext(StreamingContext)},
+   * depending on the class type.
+   */
+  private Class<?> defineContext(final Type contextType, String name, InputStream byteCodeStream) throws IOException {
+    ClassReader cr = new ClassReader(byteCodeStream);
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+
+      @Override
+      public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+        // Call super so that the method signature is registered with the ClassWriter (parent)
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+        // We only attempt to rewrite constructor
+        if (!"<init>".equals(name)) {
+          return mv;
+        }
+
+        return new AdviceAdapter(Opcodes.ASM5, mv, access, name, desc) {
+
+          boolean calledThis;
+
+          @Override
+          public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
+            // See if in this constructor it is calling other constructor (this(..)).
+            calledThis |= (opcode == Opcodes.INVOKESPECIAL
```

Comment:
Bitwise OR is really not required over here since RHS is already evaluates to a boolean, right?